src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs

```diff
@@ -0,0 +1,230 @@
+
+                // Then get the combined change
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var newText = text.WithChanges(builder);
+
+                return CompletionChange.Create(Utilities.Collapse(newText, builder.ToImmutableAndFree()));
+            }
+
+            async Task<bool> ShouldCompleteWithFullyQualifyTypeName()
+            {
+                if (!IsAddingImportsSupported(document))
+                {
+                    return true;
+                }
+
+                // We might need to qualify unimported types to use them in an import directive, because they only affect members of the containing
+                // import container (e.g. namespace/class/etc. declarations).
+                //
+                // For example, `List` and `StringBuilder` both need to be fully qualified below: 
+                // 
+                //      using CollectionOfStringBuilders = System.Collections.Generic.List<System.Text.StringBuilder>;
+                //
+                // However, if we are typing in an C# using directive that is inside a nested import container (i.e. inside a namespace declaration block), 
+                // then we can add an using in the outer import container instead (this is not allowed in VB). 
+                //
+                // For example:
+                //
+                //      using System.Collections.Generic;
+                //      using System.Text;
+                //
+                //      namespace Foo
+                //      {
+                //          using CollectionOfStringBuilders = List<StringBuilder>;
+                //      }
+                //
+                // Here we will always choose to qualify the unimported type, just to be consistent and keeps things simple.
+                return await IsInImportsDirectiveAsync(document, completionListSpan.Start, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected static bool IsAddingImportsSupported(Document document)
+        {
+            var workspace = document.Project.Solution.Workspace;
+
+            // Certain types of workspace don't support document change, e.g. DebuggerIntellisense
+            if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
+            {
+                return false;
+            }
+
+            // During an EnC session, adding import is not supported.
+            var encService = workspace.Services.GetService<IEditAndContinueWorkspaceService>();
+            if (encService?.IsDebuggingSessionInProgress == true)
+            {
+                return false;
+            }
+
+            // Certain documents, e.g. Razor document, don't support adding imports
+            var documentSupportsFeatureService = workspace.Services.GetRequiredService<IDocumentSupportsFeatureService>();
+            if (!documentSupportsFeatureService.SupportsRefactorings(document))
```

Comment:
not getting the connction between "Refactorings" and "adding imports".  we shoudl have something like `IDocumentSupportsFeatureService.SupportsAddingImports` instead to make it totally clear what we're asking about.

Reply:
this still applies.

Reply:
FYI @dibarbet 
Not gonna address in this PR though.

Reply:
#resolved.